### ISSUE TITLE
Separate configuration options for each service

### ISSUE THREAD

**Issue by kaifeng** [keylime/keylime#441]
Currently each service has their own section in the configuration file, but used without limit, like registrar accesses verifier's options, agent accesses registrar's options, tenant accesses sections of other services. This cause difficulties when we deploy the service under proxy like load balancer, because the public endpoint of a service is different from the backend address.
It could be resolved by having each service only reads its own section, [general] is a global section so it can be used for common options applied to every service.

**Comment by lukehinds** [comment]
As in the address to contact the registrar maybe different to the address the registrar binds to (so we may want two values) so we split it out?

registrar.conf
`listen_ip = 127.0.0.1`  # registrar bind address

verifier.conf
`registar_ip = 192.168.0.1` # registrar endpoint address

**Comment by kaifeng** [comment]
> 
> 
> As in the address to contact the registrar maybe different to the address the registrar binds to (so we may want two values) so we split it out?
> 
> registrar.conf
> `listen_ip = 127.0.0.1` # registrar bind address
> 
> verifier.conf
> `registar_ip = 192.168.0.1` # registrar endpoint address

Yes, we could use different configuration files but this has more impact, so a better way is to add missing options in each section.

**Comment by lukehinds** [comment]
understood, this is good proposal. 

I am open to having different configs, but I do see your point, it would make it more complicated change as we would need to fix occurances of configparser calling and redo the setup.py

**Comment by maugustosilva** [comment]
+1 on this proposal. The current configuration system quickly becomes cumbersome when dealing with more "exotic" configurations (e.g., multiple verifiers, load balancers, external databases). I ended up writing a rather complex python configurator to deal with these redundant information on different sections (to automatically generate a configuration file), but never bothered submitting it, since it was clearly a temporary solution for a much bigger problem. @kaifeng's proposal would be an important step in fixing it.

**Comment by maugustosilva** [comment]
@lukehinds @kaifeng It might look as a pet peeve of mine, but in my experience it made the initial process of understanding the inner workings of keylime (unnecessarily) harder: some options which specific to the verifier (e.g., `tls_dir`) are not explicitly marked as so (i.e., it should be `cloudverifier_tls_dir` or `verifier_tls_dir`, while options specific to the registrar are (i.e., we have `registrar_tls_dir`). In any re-organization of the sections (and options) on the configuration file, it would be good to take some time to sanitize the names of these parameters.

**Comment by kaifeng** [comment]
The tls_dir is obscure indeed, there are couple of ways to use them, I hope we could simplify the cert configuration, which means we probably need to drop some options.
Autogeneration of configuration file is not hard if suitable tooling could be utilized, oslo.config for example, is very handy and adopted in every openstack projects, but the drawback is obvious for non-openstack projects, these libraries are tightly integrated, either we use them all or use none, which may introduce undesired dependencies. The oslo team was considering moving some of libraries to the public area instead of following the openstack release model, but just not happening afaik.

**Comment by mpeters** [comment]
I also give a big +1 to this. I've run into this problem myself even on a single machine with multiple IPs where I wanted the verifier to bind to 0.0.0.0 but that breaks anything that tries to talk to the verifier.

**Comment by mpeters** [comment]
@kaifeng I believe this is done with your PR correct?

**Comment by THS-on** [comment]
There still no full separation.
I noticed that for the verifier and registrar still the following options were required because otherwise Keylime crashes:
```
[cloud_agent]
tpm_hash_alg = sha1
tpm_encryption_alg = rsa
tpm_signing_alg = rsassa
ek_handle = generate

[tenant]
tpm_cert_store = /var/lib/keylime/tpm_cert_store/
```
The changes proposed I proposed in https://gist.github.com/THS-on/4229ba2c8c83dc5d9726a6c62c932868 include moving away from the TPM abstraction for quote validation. This will probably also reduce the cross dependencies of settings because we remove shared modules between components where each component only uses a small part from.
